File: js/aspect_ratio_master.js
// Developed by 3R3BOS
import { app } from "../../scripts/app.js";

app.registerExtension({
    name: "3R3BOS.AspectRatioMaster",
    async beforeRegisterNodeDef(nodeType, nodeData, app) {
        if (nodeData.name !== "Aspect Ratio Master (3R3BOS)") return;

        const PRESETS = {
            "Low Quality": {
                video_qhd: 0.409600,
                video_ltx: 0.262144,
                video_dvd: 0.345600,
                image_eco: 0.600000,
                image_hunyuan: 1.048576,
                image_pony: 0.640000,
                image_illustrious: 1.048576,
                legacy: 0.262144
            },
            "Middle Quality": {
                video_qhd: 0.921600,
                video_ltx: 0.856064,
                video_dvd: 0.921600,
                image_eco: 1.048576,
                image_hunyuan: 2.097152,
                image_pony: 1.048576,
                image_illustrious: 1.500000,
                legacy: 0.393216
            },
            "High Quality": {
                video_qhd: 2.073600,
                video_ltx: 2.073600,
                video_dvd: 2.073600,
                image_eco: 2.097152,
                image_hunyuan: 4.194304,
                image_pony: 1.500000,
                image_illustrious: 2.359296,
                legacy: 0.640000
            }
        };

        const ARCHITECTURES = {
            "Wan Video (V2.x/V2.5)":          { round: 16, type: "video_qhd", vae_factor: 8 },
            "NVIDIA Cosmos (Gen/Predict)":   { round: 16, type: "video_dvd", vae_factor: 16 },
            "Hunyuan Video (V1/V2/V1.5)":    { round: 16, type: "video_qhd", vae_factor: 8 },
            "Mochi 1 (HD/Standard)":         { round: 16, type: "video_qhd", vae_factor: 8 },
            "HiDream (Video/Dynamic)":       { round: 16, type: "video_qhd", vae_factor: 8 },
            "LTX Video / LTX-2":             { round: 32, type: "video_ltx", vae_factor: 8 },
            "CogVideoX":                     { round: 32, type: "video_dvd", vae_factor: 8 },

            "Flux.1 / Flux.2 [Klein]":       { round: 16, type: "image_eco", vae_factor: 8 },
            "Z-Image Family (Base/Turbo/Edit)": { round: 16, type: "image_eco", vae_factor: 8 },
            "SANA-1.5 / SANA-Video":         { round: 32, type: "image_eco", vae_factor: 32 },
            "OmniGen / OmniGen2":            { round: 16, type: "image_eco", vae_factor: 8 },
            "Krea 2 (Turbo/Raw)":            { round: 16, type: "image_eco", vae_factor: 8 },
            "Aura Flow":                     { round: 16, type: "image_eco", vae_factor: 8 },
            "Lumina / Lumina-Next":          { round: 32, type: "image_eco", vae_factor: 8 },
            "PixArt Alpha / Sigma":          { round: 32, type: "image_eco", vae_factor: 8 },
            "Hunyuan Image 1.0":             { round: 16, type: "image_eco", vae_factor: 8 },
            "Hunyuan Image 2.1 (2K)":        { round: 16, type: "image_hunyuan", vae_factor: 8 },

            "SDXL (Base/Turbo/Lightning)":   { round: 8,  type: "image_pony", vae_factor: 8 },
            "Pony / Pony V7":                { round: 8,  type: "image_pony", vae_factor: 8 },
            "Kolors (Kwai)":                 { round: 8,  type: "image_pony", vae_factor: 8 },
            "Illustrious / NoobAI":          { round: 8,  type: "image_illustrious", vae_factor: 8 },

            "SD 3.5 (Large/Turbo)":          { round: 64, type: "image_eco", vae_factor: 8 },
            "SD 3.0":                        { round: 16, type: "image_eco", vae_factor: 8 },
            "SD 1.5 / LCM":                  { round: 8,  type: "legacy", vae_factor: 8 }
        };

        function getRes(arch, mode, ratioStr, scale, useCustomRatio = false, customW = 1024, customH = 1024) {
            const conf = ARCHITECTURES[arch] || ARCHITECTURES["Wan Video (V2.x/V2.5)"];
            const r = conf.round;
            const archType = conf.type;

            let mKey = "Middle Quality";
            if (mode.includes("Low")) mKey = "Low Quality";
            else if (mode.includes("High")) mKey = "High Quality";

            const presetGroup = PRESETS[mKey];
            const mp = presetGroup[archType] || 1.0;

            let w, h;
            if (useCustomRatio) {
                w = Math.round((customW * scale) / r) * r;
                h = Math.round((customH * scale) / r) * r;
                return { w, h, mode: mKey, conf };
            }

            const total = mp * (scale * scale) * 1000000;
            const p = ratioStr.split(':');
            let rw = parseFloat(p[0]);
            let rh = parseFloat(p[1]);

            if (isNaN(rw) || isNaN(rh)) { rw = 1.0; rh = 1.0; }

            const s = Math.sqrt(total / (rw * rh));
            w = Math.round((s * rw) / r) * r;
            h = Math.round((s * rh) / r) * r;

            return { w, h, mode: mKey, conf };
        }

        function drawMiniRatioShape(ctx, x, y, maxW, maxH, ratioStr, highlight, activeColor) {
            const p = ratioStr.split(':');
            let rw = parseFloat(p[0]);
            let rh = parseFloat(p[1]);
            if (isNaN(rw) || isNaN(rh)) return;

            let w = maxW; let h = maxH;
            const rVal = rw / rh;
            if (rVal > 1) {
                h = maxW / rVal;
                if (h > maxH) { h = maxH; w = maxH * rVal; }
            } else {
                w = maxH * rVal;
                if (w > maxW) { w = maxW; h = maxW / rVal; }
            }

            const bx = x - w / 2; const by = y - h / 2;
            ctx.beginPath();
            ctx.roundRect(bx, by, w, h, 2);
            ctx.strokeStyle = highlight ? activeColor : "rgba(255, 255, 255, 0.25)";
            ctx.lineWidth = 1;
            ctx.stroke();

            if (highlight) {
                ctx.fillStyle = activeColor + "22";
                ctx.fill();
            }
        }

        if (!nodeType.prototype.onNodeCreated.__patched) {
            const onNodeCreated = nodeType.prototype.onNodeCreated;
            nodeType.prototype.onNodeCreated = function () {
                if (onNodeCreated) onNodeCreated.apply(this, arguments);
                this.setSize([380, 680]);
                this.min_size = [380, 680];
                this.color = "#242730";
                this.bgcolor = "#1b1e24";

                const refresh = () => { this.setDirtyCanvas(true, true); };
                this.widgets.forEach(w => {
                    const cb = w.callback;
                    w.callback = (v) => { if (cb) cb(v); refresh(); };
                });

                if (!this._animTimer) {
                    this._animTimer = setInterval(() => {
                        if (this.flags && !this.flags.collapsed && app.canvas.visible_nodes && app.canvas.visible_nodes.includes(this)) {
                            this.setDirtyCanvas(true, true);
                        }
                    }, 60);
                }
            };
            nodeType.prototype.onNodeCreated.__patched = true;
        }

        nodeType.prototype.onDrawForeground = function (ctx) {
            if (this.flags.collapsed) return;

            if (this.size[0] < 380 || this.size[1] < 680) {
                this.size[0] = Math.max(this.size[0], 380);
                this.size[1] = Math.max(this.size[1], 680);
            }

            const wCustomToggle = this.widgets.find(w => w.name === "use_custom_ratio");
            const wCustomW = this.widgets.find(w => w.name === "custom_width");
            const wCustomH = this.widgets.find(w => w.name === "custom_height");
            const wArch = this.widgets.find(w => w.name === "model_arch");
            const wMode = this.widgets.find(w => w.name === "performance_mode");
            const wRatio = this.widgets.find(w => w.name === "aspect_ratio");
            const wScale = this.widgets.find(w => w.name === "custom_scale_factor");

            const arch = wArch ? wArch.value : "Wan Video (V2.x/V2.5)";
            const mode = wMode ? wMode.value : "Middle Quality";
            const curRatio = wRatio ? wRatio.value : "16:9 (Cinematic)";
            const useCustomRatio = wCustomToggle ? wCustomToggle.value : false;
            const customW = wCustomW ? wCustomW.value : 1024;
            const customH = wCustomH ? wCustomH.value : 1024;
            const scale = wScale ? wScale.value : 1.0;

            const margin = 10;
            const topY = 350;
            const flipH = 28;
            const gw = this.size[0] - margin * 2;

            const currentRes = getRes(arch, mode, curRatio, scale, useCustomRatio, customW, customH);

            this.color = "#242730";
            this.bgcolor = "#1b1e24";

            const accentCol = "#00b4d8";
            const accentBg = "rgba(0, 180, 216, 0.12)";
            const accentBgHover = "rgba(0, 180, 216, 0.22)";
            const pulse = Math.sin(Date.now() / 300) * 0.35 + 0.65;

            ctx.beginPath();
            ctx.roundRect(margin, topY, gw, flipH, 6);

            if (this.hoverFlip) {
                ctx.fillStyle = accentBgHover; ctx.strokeStyle = accentCol; ctx.lineWidth = 1.5;
            } else {
                ctx.fillStyle = "rgba(20, 20, 24, 0.6)"; ctx.strokeStyle = "rgba(255, 255, 255, 0.05)"; ctx.lineWidth = 1;
            }
            ctx.fill(); ctx.stroke();

            ctx.fillStyle = this.hoverFlip ? "#fff" : "#ccc";
            ctx.font = "bold 11px Arial"; ctx.textAlign = "center";
            ctx.fillText(`Flip Aspect Ratio 🔄`, margin + gw / 2, topY + 18);

            this.flipBtnBox = { l: margin, t: topY, r: margin + gw, b: topY + flipH };

            const RATIOS = [
                { l: "1:1", v: "1:1" }, { l: "4:3", v: "4:3" }, { l: "3:4", v: "3:4" },
                { l: "16:9", v: "16:9" }, { l: "9:16", v: "9:16" }, { l: "21:9", v: "21:9" },
                { l: "9:21", v: "9:21" }, { l: "3:2", v: "3:2" }, { l: "2:3", v: "2:3" },
                { l: "2:1", v: "2:1" }, { l: "1:2", v: "1:2" }, { l: "5:4", v: "5:4" }
            ];

            const gridY = topY + flipH + 8;
            const bottomBarH = 26;
            const gh = this.size[1] - gridY - bottomBarH - 15;
            const ch = gh / 4;
            const cw = gw / 3;

            this.hitBoxes = [];
            ctx.save();
            ctx.translate(margin, gridY);

            RATIOS.forEach((r, i) => {
                const col = i % 3;
                const row = Math.floor(i / 3);
                const x = col * cw;
                const y = row * ch;

                // FIX: On force à false ici pour garder l'aperçu du ratio propre sur la grille
                const res = getRes(arch, mode, r.v, scale, false, customW, customH);
                const active = !useCustomRatio && curRatio.startsWith(r.v);
                const hovered = this.hoveredIndex === i;

                this.hitBoxes.push({ val: r.v, l: margin + x, t: gridY + y, r: margin + x + cw, b: gridY + y + ch });

                ctx.save();
                ctx.beginPath();
                ctx.roundRect(x + 2, y + 2, cw - 4, ch - 4, 6);

                if (active) {
                    ctx.fillStyle = accentBg; ctx.strokeStyle = accentCol; ctx.lineWidth = 1.5;
                    ctx.shadowColor = accentCol; ctx.shadowBlur = 3 + pulse * 5;
                } else if (hovered) {
                    ctx.fillStyle = "rgba(255, 255, 255, 0.08)"; ctx.strokeStyle = "rgba(255, 255, 255, 0.25)"; ctx.lineWidth = 1;
                } else {
                    ctx.fillStyle = "rgba(255, 255, 255, 0.02)"; ctx.strokeStyle = "rgba(255, 255, 255, 0.06)"; ctx.lineWidth = 1;
                }
                ctx.fill(); ctx.stroke();
                ctx.restore();

                const cardCy = y + ch / 2;
                const shapeX = x + 22;
                drawMiniRatioShape(ctx, shapeX, cardCy, 28, 28, r.v, active || hovered, accentCol);

                const textStartX = x + 44;
                ctx.textAlign = "left";
                ctx.fillStyle = active ? "#ffffff" : (hovered ? "#ffffff" : "#bbbbbb");
                ctx.font = active ? "bold 11px Arial" : "11px Arial";
                ctx.fillText(r.l, textStartX, cardCy - 4);

                ctx.fillStyle = active ? accentCol : (hovered ? "#aaaaaa" : "#777777");
                ctx.font = "9px Consolas";
                ctx.fillText(`${res.w}x${res.h}`, textStartX, cardCy + 8);
            });
            ctx.restore();

            const barY = this.size[1] - bottomBarH - 8;
            ctx.beginPath();
            ctx.roundRect(margin, barY, gw, bottomBarH, 4);
            ctx.fillStyle = "rgba(10, 10, 12, 0.7)"; ctx.strokeStyle = "rgba(255,255,255,0.05)"; ctx.lineWidth = 1;
            ctx.fill(); ctx.stroke();

            const vaeF = currentRes.conf.vae_factor || 8;
            const latentW = Math.floor(currentRes.w / vaeF);
            const latentH = Math.floor(currentRes.h / vaeF);
            const mpCount = (currentRes.w * currentRes.h / 1000000).toFixed(2);

            ctx.fillStyle = accentCol; ctx.font = "bold 10px Consolas"; ctx.textAlign = "center";
            ctx.fillText(`${currentRes.w}x${currentRes.h} (${mpCount}MP) | Latents: ${latentW}x${latentH}`, margin + gw / 2, barY + 16);
        };

        nodeType.prototype.onMouseMove = function (e, pos) {
            const x = pos[0]; const y = pos[1];
            let hoveredIndex = -1; let hoverFlip = false;

            if (x >= 0 && x <= this.size[0] && y >= 0 && y <= this.size[1]) {
                if (this.hitBoxes) {
                    for (let i = 0; i < this.hitBoxes.length; i++) {
                        const h = this.hitBoxes[i];
                        if (x >= h.l && x <= h.r && y >= h.t && y <= h.b) { hoveredIndex = i; break; }
                    }
                }
                if (this.flipBtnBox) {
                    const fb = this.flipBtnBox;
                    if (x >= fb.l && x <= fb.r && y >= fb.t && y <= fb.b) { hoverFlip = true; }
                }
            }

            if (this.hoveredIndex !== hoveredIndex || this.hoverFlip !== hoverFlip) {
                this.hoveredIndex = hoveredIndex; this.hoverFlip = hoverFlip;
                this.setDirtyCanvas(true, true);
            }
        };

        nodeType.prototype.onMouseLeave = function () {
            this.hoveredIndex = -1; this.hoverFlip = false;
            this.setDirtyCanvas(true, true);
        };

        nodeType.prototype.onMouseDown = function (e, pos) {
            if (e.buttons !== 1) return false;
            const x = pos[0]; const y = pos[1];

            if (this.flipBtnBox) {
                const fb = this.flipBtnBox;
                if (x >= fb.l && x <= fb.r && y >= fb.t && y <= fb.b) {
                    const wRatio = this.widgets.find(w => w.name === "aspect_ratio");
                    const wCustomToggle = this.widgets.find(w => w.name === "use_custom_ratio");
                    const useCustomRatio = wCustomToggle ? wCustomToggle.value : false;

                    if (useCustomRatio) {
                        const wCustomW = this.widgets.find(w => w.name === "custom_width");
                        const wCustomH = this.widgets.find(w => w.name === "custom_height");
                        if (wCustomW && wCustomH) {
                            const valW = wCustomW.value; const valH = wCustomH.value;
                            wCustomW.value = valH; wCustomH.value = valW;
                            if (wCustomW.callback) wCustomW.callback(valH);
                            if (wCustomH.callback) wCustomH.callback(valW);
                        }
                    } else if (wRatio) {
                        const curVal = wRatio.value;
                        const lookup = {
                            "16:9 (Cinematic)": "9:16 (Stories)",
                            "9:16 (Stories)": "16:9 (Cinematic)",
                            "4:3 (Photo)": "3:4 (Portrait)",
                            "3:4 (Portrait)": "4:3 (Photo)",
                            "21:9 (Ultrawide)": "9:21 (Tower)",
                            "9:21 (Tower)": "21:9 (Ultrawide)",
                            "3:2 (DSLR)": "2:3 (Classic)",
                            "2:3 (Classic)": "3:2 (DSLR)",
                            "5:4 (Medium)": "4:5 (Social)",
                            "4:5 (Social)": "5:4 (Medium)",
                            "2:1 (Univisium)": "1:2 (Portrait Extra)",
                            "1:2 (Portrait Extra)": "2:1 (Univisium)"
                        };
                        const nextVal = lookup[curVal];
                        if (nextVal && wRatio.options.values.includes(nextVal)) {
                            wRatio.value = nextVal;
                            if (wRatio.callback) wRatio.callback(nextVal);
                        }
                    }
                    this.setDirtyCanvas(true, true);
                    return true;
                }
            }

            if (this.hitBoxes) {
                for (const h of this.hitBoxes) {
                    if (x >= h.l && x <= h.r && y >= h.t && y <= h.b) {
                        const wRatio = this.widgets.find(w => w.name === "aspect_ratio");
                        const wCustomToggle = this.widgets.find(w => w.name === "use_custom_ratio");
                        if (wRatio) {
                            const opt = wRatio.options.values.find(v => v.startsWith(h.val));
                            if (opt) {
                                if (wCustomToggle) {
                                    wCustomToggle.value = false;
                                    if (wCustomToggle.callback) wCustomToggle.callback(false);
                                }
                                wRatio.value = opt;
                                if (wRatio.callback) wRatio.callback(opt);
                            }
                        }
                        return true;
                    }
                }
            }
        };

        nodeType.prototype.onRemoved = function() {
            if (this._animTimer) clearInterval(this._animTimer);
        };
    }
});

File: nodes/aspect_ratio_master.py
# Developed by 3R3BOS
import torch
import numpy as np
from PIL import Image, ImageDraw, ImageFont

PRESETS = {
    "Low Quality": {
        "video_qhd": 0.409600,
        "video_ltx": 0.262144,
        "video_dvd": 0.345600,
        "image_eco": 0.600000,
        "image_hunyuan": 1.048576,
        "image_pony": 0.640000,
        "image_illustrious": 1.048576,
        "legacy": 0.262144
    },
    "Middle Quality": {
        "video_qhd": 0.921600,
        "video_ltx": 0.856064,
        "video_dvd": 0.921600,
        "image_eco": 1.048576,
        "image_hunyuan": 2.097152,
        "image_pony": 1.048576,
        "image_illustrious": 1.500000,
        "legacy": 0.393216
    },
    "High Quality": {
        "video_qhd": 2.073600,
        "video_ltx": 2.073600,
        "video_dvd": 2.073600,
        "image_eco": 2.097152,
        "image_hunyuan": 4.194304,
        "image_pony": 1.500000,
        "image_illustrious": 2.359296,
        "legacy": 0.640000
    }
}

ARCHITECTURES = {
    # --- VIDEO ---
    # Wan 2.1/2.2: 3D causal VAE, 16 latent channels, 8x spatial
    "Wan Video (V2.x/V2.5)":          {"round": 16, "type": "video_qhd", "vae_factor": 8,  "channels": 16},
    # Cosmos: 16ch latent, 16x spatial compression
    "NVIDIA Cosmos (Gen/Predict)":   {"round": 16, "type": "video_dvd", "vae_factor": 16, "channels": 16},
    # Hunyuan Video: 3D causal VAE, 16ch
    "Hunyuan Video (V1/V2/V1.5)":    {"round": 16, "type": "video_qhd", "vae_factor": 8,  "channels": 16},
    # Mochi 1: 3D VAE with 12 latent channels, 8x spatial, 6x temporal
    "Mochi 1 (HD/Standard)":         {"round": 16, "type": "video_qhd", "vae_factor": 8,  "channels": 12},
    # HiDream video: uses Flux-compatible 16ch VAE
    "HiDream (Video/Dynamic)":       {"round": 16, "type": "video_qhd", "vae_factor": 8,  "channels": 16},
    # LTX Video: unique 128ch latent, 32x spatial (DC-VAE)
    "LTX Video / LTX-2":             {"round": 32, "type": "video_ltx", "vae_factor": 32, "channels": 128},
    # CogVideoX: 3D causal VAE, 16ch
    "CogVideoX":                     {"round": 32, "type": "video_dvd", "vae_factor": 8,  "channels": 16},

    # --- NEXT-GEN IMAGE ---
    # Flux: 16ch, 8x spatial
    "Flux.1 / Flux.2 [Klein]":       {"round": 16, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    # Z-Image: 16ch (same family as Flux)
    "Z-Image Family (Base/Turbo/Edit)": {"round": 16, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    # SANA: DC-AE f32c32, 32ch latent, 32x spatial
    "SANA-1.5 / SANA-Video":         {"round": 32, "type": "image_eco", "vae_factor": 32, "channels": 32},
    # OmniGen2: 16ch latent diffusion path
    "OmniGen / OmniGen2":            {"round": 16, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    # Krea 2: Qwen-Image VAE, 16ch
    "Krea 2 (Turbo/Raw)":            {"round": 16, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    # AuraFlow: 16ch latent space
    "Aura Flow":                     {"round": 16, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    # Lumina/Lumina-Next: 16ch
    "Lumina / Lumina-Next":          {"round": 32, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    # PixArt Alpha/Sigma: uses SDXL-compatible 4ch VAE
    "PixArt Alpha / Sigma":          {"round": 32, "type": "image_eco", "vae_factor": 8,  "channels": 4},
    # Hunyuan Image 1.0: 16ch, 8x
    "Hunyuan Image 1.0":             {"round": 16, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    # Hunyuan Image 2.1: 16ch, 32x spatial compression
    "Hunyuan Image 2.1 (2K)":        {"round": 16, "type": "image_hunyuan", "vae_factor": 32, "channels": 16},

    # --- SDXL / PONY ---
    "SDXL (Base/Turbo/Lightning)":   {"round": 8,  "type": "image_pony", "vae_factor": 8,  "channels": 4},
    "Pony / Pony V7":                {"round": 8,  "type": "image_pony", "vae_factor": 8,  "channels": 4},
    "Kolors (Kwai)":                 {"round": 8,  "type": "image_pony", "vae_factor": 8,  "channels": 4},
    "Illustrious / NoobAI":          {"round": 8,  "type": "image_illustrious", "vae_factor": 8,  "channels": 4},

    # --- LEGACY ---
    # SD3.x: 16ch VAE
    "SD 3.5 (Large/Turbo)":          {"round": 64, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    "SD 3.0":                        {"round": 16, "type": "image_eco", "vae_factor": 8,  "channels": 16},
    "SD 1.5 / LCM":                  {"round": 8,  "type": "legacy",    "vae_factor": 8,  "channels": 4}
}

def pil2tensor(image):
    return torch.from_numpy(np.array(image).astype(np.float32) / 255.0).unsqueeze(0)

class AspectRatioMaster:
    @classmethod
    def INPUT_TYPES(s):
        sorted_archs = sorted(list(ARCHITECTURES.keys()))
        priority = [
            "Wan Video (V2.x/V2.5)", 
            "Flux.1 / Flux.2 [Klein]", 
            "Z-Image Family (Base/Turbo/Edit)", 
            "NVIDIA Cosmos (Gen/Predict)"
        ]
        for p in reversed(priority):
            if p in sorted_archs:
                sorted_archs.insert(0, sorted_archs.pop(sorted_archs.index(p)))

        return {
            "required": {
                "use_custom_ratio": ("BOOLEAN", {"default": False, "label_on": "CUSTOM ON", "label_off": "CUSTOM OFF"}),
                "custom_width": ("INT", {"default": 1024, "min": 64, "max": 8192, "step": 8}),
                "custom_height": ("INT", {"default": 1024, "min": 64, "max": 8192, "step": 8}),
                "model_arch": (sorted_archs, {"default": "Wan Video (V2.x/V2.5)"}),
                "performance_mode": (["Low Quality", "Middle Quality", "High Quality"], {"default": "Middle Quality"}),
                "aspect_ratio": ([
                    "1:1 (Square)", "4:3 (Photo)", "3:4 (Portrait)",
                    "16:9 (Cinematic)", "9:16 (Stories)", "21:9 (Ultrawide)",
                    "9:21 (Tower)", "3:2 (DSLR)", "2:3 (Classic)",
                    "2:1 (Univisium)", "1:2 (Portrait Extra)", "5:4 (Medium)", "4:5 (Social)"
                ], {"default": "16:9 (Cinematic)"}),
                "use_batch": ("BOOLEAN", {"default": False, "label_on": "BATCH ON", "label_off": "BATCH OFF"}),
                "batch_size": ("INT", {"default": 1, "min": 1, "max": 64}),
                "custom_scale_factor": ("FLOAT", {"default": 1.0, "min": 0.5, "max": 2.0, "step": 0.05}),
            },
            "hidden": {"unique_id": "UNIQUE_ID"},
        }

    RETURN_TYPES = ("INT", "INT", "LATENT", "IMAGE")
    RETURN_NAMES = ("width", "height", "empty_latent", "preview")
    FUNCTION = "scout"
    CATEGORY = "3R3BOS/Utils"

    def get_font(self, size):
        fonts = ["arial.ttf", "segoeui.ttf", "Roboto-Regular.ttf", "DejaVuSans.ttf"]
        for font_name in fonts:
            try: return ImageFont.truetype(font_name, size)
            except IOError: continue
        return ImageFont.load_default()

    def create_preview_image(self, width, height, ratio_str, arch_name, mode_name, batch_info):
        W, H = 800, 400
        img = Image.new('RGBA', (W, H), (36, 39, 48, 255))
        draw = ImageDraw.Draw(img)

        for i in range(0, W, 40): draw.line([(i, 0), (i, H)], fill=(45, 48, 58, 255))
        for i in range(0, H, 40): draw.line([(0, i), (W, i)], fill=(45, 48, 58, 255))

        f_xl = self.get_font(60)
        f_md = self.get_font(24)
        f_sm = self.get_font(16)

        draw.rectangle([0, 0, 300, H], fill=(27, 30, 36, 255))
        draw.line([(300, 0), (300, H)], fill=(38, 42, 51, 255), width=2)

        y = 40
        def draw_label(lbl, val, col="#ffffff"):
            nonlocal y
            draw.text((20, y), lbl, fill="#8f9cae", font=f_sm)
            draw.text((20, y+20), str(val), fill=col, font=f_md)
            y += 70

        short_arch = arch_name.split('(')[0].split('/')[0].strip()
        draw_label("ARCHITECTURE", short_arch, "#00b4d8")
        draw_label("QUALITY", mode_name.upper(), "#00b4d8")
        draw_label("ASPECT RATIO", ratio_str.split(' ')[0])
        if batch_info != "SINGLE":
            draw_label("BATCH", batch_info, "#FF0055")

        cx, cy = 300 + (W-300)//2, H//2
        scale = 0.15
        bw, bh = width * scale, height * scale
        
        max_w, max_h = 450, 350
        if bw > max_w: s=max_w/bw; bw*=s; bh*=s
        if bh > max_h: s=max_h/bh; bw*=s; bh*=s

        x1, y1 = cx - bw/2, cy - bh/2
        x2, y2 = cx + bw/2, cy + bh/2

        border_col = (0, 180, 216)
        draw.rectangle([x1, y1, x2, y2], fill=(0, 180, 216, 40), outline=border_col, width=3)
        
        txt = f"{width} x {height}"
        bbox = draw.textbbox((0,0), txt, font=f_xl)
        tw, th = bbox[2]-bbox[0], bbox[3]-bbox[1]
        
        draw.text((cx - tw/2 + 2, cy - th/2 + 2), txt, fill=(0, 0, 0, 255), font=f_xl)
        draw.text((cx - tw/2, cy - th/2), txt, fill=(255, 255, 255, 255), font=f_xl)

        return pil2tensor(img.convert('RGB'))

    def scout(self, model_arch, performance_mode, aspect_ratio, use_custom_ratio, custom_width, custom_height, use_batch, batch_size, custom_scale_factor, unique_id):
        conf = ARCHITECTURES.get(model_arch, ARCHITECTURES["Wan Video (V2.x/V2.5)"])
        round_to = conf["round"]
        arch_type = conf["type"]
        vae_factor = conf.get("vae_factor", 8)
        latent_channels = conf.get("channels", 4)

        selected_preset = PRESETS.get(performance_mode, PRESETS["Middle Quality"])
        base_mp = selected_preset.get(arch_type, 1.0)
        final_mp = base_mp * (custom_scale_factor ** 2)

        if use_custom_ratio:
            w = int(round((custom_width * custom_scale_factor) / round_to) * round_to)
            h = int(round((custom_height * custom_scale_factor) / round_to) * round_to)
        else:
            rw, rh = 1.0, 1.0
            try:
                p = aspect_ratio.split(' ')[0].split(':')
                rw, rh = float(p[0]), float(p[1])
            except:
                rw, rh = 1.0, 1.0

            pixels = final_mp * 1_000_000
            s = (pixels / (rw * rh)) ** 0.5
            w = int(round(s * rw / round_to) * round_to)
            h = int(round(s * rh / round_to) * round_to)

        b_count = batch_size if use_batch else 1
        latent = torch.zeros([b_count, latent_channels, h // vae_factor, w // vae_factor])

        batch_str = str(b_count) if use_batch else "SINGLE"
        ratio_str_val = f"{w}:{h}" if use_custom_ratio else aspect_ratio
        prev = self.create_preview_image(w, h, ratio_str_val, model_arch, performance_mode, batch_str)

        return (w, h, {"samples": latent}, prev)
